### ISSUE TITLE
Loading Chunks from OS to Read Buffer is slow

### ISSUE THREAD

**Issue by e-dard** [influxdata/influxdb_iox#3321]
**TLDR**; loading parquet into the Read Buffer via the InfluxDB load chunks CLI  command is 10–25x slower than loading the parquet into the Read Buffer directly, using a Parquet reader to extract record batches.

Recently I have been loading chunks (stored as Parquet files) back into the read buffer using the CLI added in #3250 and #3268.

I noticed that the process seemed incredibly slow on my local machine (Mac M1 8 core / 16GB memory). I was running IOx in `release` mode.

To verify that something was up, I adapted a program I keep on a branch that I use to load parquet into the Read Buffer for development/testing purposes. Then I ran two experiments.

## Loading Chunks via CLI in IOx

On my machine I have 13 parquet files that belong to a database. In total they're about 500MB containing around 7M rows of data.

I compiled IOx as follows: `RUSTFLAGS="-C target-cpu=native" cargo build --release`. I then started IOx up like this:

```
INFLUXDB_IOX_ID=100 ./target/release/influxdb_iox run --object-store=file --data-dir=/Users/edd/Documents/flux_query_log/iox_parquet
```

and verified the parquet files were located in the Object Store (which is a  local path on my machine):

```
target/debug/influxdb_iox database query query_log "select id, storage  from system.chunks"
+--------------------------------------+-----------------+
| id                                   | storage         |
+--------------------------------------+-----------------+
| fb425523-c822-486c-b317-72a3ba7dbe57 | ObjectStoreOnly |
| 48642d71-1206-42a6-a5d8-5bc0faa752d4 | ObjectStoreOnly |
| 012e9c79-65c8-419c-8a60-c26fbe0bb913 | ObjectStoreOnly |
| 47f7d3a6-714e-4ffc-b8f6-0d144c7653d8 | ObjectStoreOnly |
| 5fafd0c1-0d0d-4798-87a0-6accb7d54ec1 | ObjectStoreOnly |
| 33837661-7501-460a-9b66-e64549b25b1c | ObjectStoreOnly |
| d2071087-31ab-4e69-90b8-b4e9d81c97fb | ObjectStoreOnly |
| 9b8985bb-0a1e-4c85-b4df-564d9e43f079 | ObjectStoreOnly |
| dddb3c40-15ca-4c02-9360-d865094ee092 | ObjectStoreOnly |
| 91be1c79-d7da-41f0-b52f-f8816fe1bc7c | ObjectStoreOnly |
| 4e265d94-1479-43f4-8699-7b35738c0d30 | ObjectStoreOnly |
| 2612a9fe-e823-432c-bc65-5c840fd21a08 | ObjectStoreOnly |
| b2152255-5ca8-455d-bbc2-59cdbfbf6818 | ObjectStoreOnly |
+--------------------------------------+-----------------+
```

Ok, next we will load the chunks into the Read Buffer and time how long it takes. For reasons of expediency I periodically ran the above query until all the chunks were seen to be in the Read Buffer.

To kick off the chunk loading:

```
xargs -I % target/release/influxdb_iox database chunk load query_log % < chunk_ids
```

(where `chunk_ids`) is a list of the IDs in the above query.

Whilst this was running I took a CPU profile. I noted that time was being spent in `SortStream`, which is a physical plan for sorting record batch data. This seems like a prime candidate for part of the problem - there is no need to sort persisted data before putting it into the Read Buffer as it should already be sorted.

![Screenshot 2021-12-07 at 11 52 02](https://user-images.githubusercontent.com/501993/145024337-b4dfc5aa-60bd-469c-85e7-f5f356fb1e52.png)

In total the time it took was **3m20s**.

## Loading Chunks directly into Read Buffer

On `er/parquet_load_rub_example` there is a simple program that loads chunks into the Read Buffer from Parquet files. It's short enough to list here verbatim:

```rust
use datafusion::parquet::arrow::{ArrowReader, ParquetFileArrowReader};
use datafusion::parquet::file::serialized_reader::SerializedFileReader;
use read_buffer::RBChunk;
use std::fs::File;
use std::sync::mpsc::channel;
use std::sync::Arc;
use threadpool::ThreadPool;

fn main() {
    let files = vec![
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 06:00:00/9b8985bb-0a1e-4c85-b4df-564d9e43f079.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 06:00:00/dddb3c40-15ca-4c02-9360-d865094ee092.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 05:00:00/33837661-7501-460a-9b66-e64549b25b1c.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 07:00:00/48642d71-1206-42a6-a5d8-5bc0faa752d4.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 07:00:00/fb425523-c822-486c-b317-72a3ba7dbe57.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 07:00:00/012e9c79-65c8-419c-8a60-c26fbe0bb913.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 10:00:00/5fafd0c1-0d0d-4798-87a0-6accb7d54ec1.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 10:00:00/47f7d3a6-714e-4ffc-b8f6-0d144c7653d8.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 09:00:00/4e265d94-1479-43f4-8699-7b35738c0d30.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 09:00:00/91be1c79-d7da-41f0-b52f-f8816fe1bc7c.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 11:00:00/d2071087-31ab-4e69-90b8-b4e9d81c97fb.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 08:00:00/b2152255-5ca8-455d-bbc2-59cdbfbf6818.parquet",
        "/Users/edd/Documents/flux_query_log/iox_parquet/dbs/925280ba-4f8d-413a-a955-d86dd4b172d9/data/query_log/2021-11-30 08:00:00/2612a9fe-e823-432c-bc65-5c840fd21a08.parquet",
    ];

    let batch_size = 1024 * 25; // matches current batch size in IOx

    let num_threads = num_cpus::get();
    // let num_threads = 13;

    let now = std::time::Instant::now();
    println!("creating thread pool of size {:?}", num_threads);
    let pool = ThreadPool::new(num_threads);
    let (tx, rx) = channel();

    for file in files.clone() {
        let tx = tx.clone();
        pool.execute(move || {
            println!("creating RUB chunk");
            let now = std::time::Instant::now();
            let chunk = load_chunk_rub(file, batch_size);

            let elapsed = now.elapsed().as_secs();

            let throughput = if elapsed == 0 {
                f64::INFINITY
            } else {
                chunk.rows() as f64 / elapsed as f64
            };

            println!(
                "RUB chunk created in {:?} {:?} rows/sec",
                now.elapsed(),
                throughput
            );

            tx.send(chunk.row_groups()).expect("Could not send data!");
        });
    }

    let mut row_groups = 0;
    for _ in &files {
        row_groups += rx.recv().unwrap();
    }

    println!(
        "Finished in {:?}. Loaded {:?} chunks with {:?} row groups to the Read Buffer",
        now.elapsed(),
        files.len(),
        row_groups,
    );
}

fn load_chunk_rub(path: &str, batch_size: usize) -> RBChunk {
    println!("loading parquet file from {:?}", path);
    let file = File::open(path).unwrap();
    let file_reader = SerializedFileReader::new(file).unwrap();
    let mut arrow_reader = ParquetFileArrowReader::new(Arc::new(file_reader));

    let record_batch_reader = arrow_reader.get_record_reader(batch_size).unwrap();
    let mut itr = record_batch_reader;
    let cm = read_buffer::ChunkMetrics::new_unregistered();

    let mut chunk = read_buffer::RBChunk::new("my_table", itr.next().unwrap().unwrap(), cm);
    for rb in itr {
        chunk.upsert_table(rb.unwrap());
    }

    chunk
}
```

Effectively it creates a thread pool and then calls the usual Read Buffer loading code, being fed record batches read directly from the Parquet files.

This program was built as follows:

```
RUSTFLAGS="-C target-cpu=native" cargo build --release -p read_buffer
```

I ran several different threadpool sizes.

 - 1T: single threaded loading. Loads one parquet at a time;
 - 8T: a single thread per HW thread on my machine. New parquet loads added as threads get freed up.
 - 13T: more threads in pool than HW theads on my machine. OS responsible for scheduling them across 8 HW threads.

## Results

Results are below. As you would expect, pairing the amount of work to the number of HW threads results in the best performance. 

Currently, using the IOx CLI is 10–25x slower. I suspect due in part to unnecessary sorting, but also I wonder due to far too much asynchronous yielding and context switching. The latter thought needs more investigation though.

| IOx Load CLI  | RUB Loader 1T   | RUB Loader 8T   | RUB 1Loader 3T    |
|---|---|---|---|
| 200s  |  19s  | 7.5s   |  13s  |

**Comment by e-dard** [comment]
This is the loading method the CLI uses:

https://github.com/influxdata/influxdb_iox/blob/f0dcc6b1b014273e07190f3ec231cb03563d7900/server/src/db/lifecycle/load.rs#L20-L75

It looks to be using the `compact_plan`, which does de-duplication, sorting etc. I don't think we should be doing any of that to load a chunk back into the RUB.

**Comment by e-dard** [comment]
Here is the current plan being run:

```
2021-12-07T12:46:25.358003Z DEBUG query::exec::context: prepare_plan: plan to run text=SortPreservingMergeExec: [env@0 ASC,error@1 ASC,errorCode@2 ASC,errorType@3 ASC,hostname@4 ASC,orgID@5 ASC,ot_trace_sampled@6 ASC,result@8 ASC,table@9 ASC,time@10 ASC]
  RepartitionExec: partitioning=RoundRobinBatch(8)
    SortExec: [env@0 ASC,error@1 ASC,errorCode@2 ASC,errorType@3 ASC,hostname@4 ASC,orgID@5 ASC,ot_trace_sampled@6 ASC,result@8 ASC,table@9 ASC,time@10 ASC]
      IOxReadFilterNode: table_name=query_log, chunks=1 predicate=Predicate
```

**Comment by e-dard** [comment]
Fix will be to  avoid the sort in `sorted_scan_plan`

**Comment by NGA-TRAN** [comment]
Very nice discovery @e-dard . This is alway something we run into when we add a new feature (load OS back to RUB) using available plan that was not optimize for it :)

**Comment by NGA-TRAN** [comment]
I think we may want to think about a general way to always verify if the output of the operator is already sorted to avoid extra sort for all future new features, too.
